d-sandbox

<div style="text-align: center; line-height: 0; padding-top: 9px;">
  <img src="https://databricks.com/wp-content/uploads/2018/03/db-academy-rgb-1200px.png" alt="Databricks Learning" style="width: 600px">
</div>

# Hyperparameter Search Lab

**Objective**: *Apply grid-search hyperparameter optimization to improve the performance of a model.*

In this lab, you will apply what you've learned in this lesson. When complete, please use the answers to the exercises to answer questions in the following quiz within Coursera.

In [0]:
%run "../../Includes/Classroom-Setup"

Mounting course-specific datasets to /mnt/training ... Datasets are already mounted to /mnt/training from s3a://databricks-corp-training/common

Out[2]: DataFrame[]

res1: Boolean = false

res2: Boolean = false

res3: Boolean = false

-sandbox

## Exercise 1

In this exercise, you will create an enhanced user-level table to try to better predict whether or not each user takes at least 10,000 steps in a day.

Fill in the blanks in the below cell to create the `adsda.ht_user_metrics_hs_lab` table.

<img alt="Hint" title="Hint" style="vertical-align: text-bottom; position: relative; height:1.75em; top:0.3em" src="https://files.training.databricks.com/static/images/icon-light-bulb.svg"/>&nbsp;**Hint:** Refer back to previous demos on how to create the `steps_10000` column.

In [0]:
%sql
-- ANSWER
CREATE OR REPLACE TABLE adsda.ht_user_metrics_hs_lab
USING DELTA LOCATION "/adsda/ht-user-metrics-hs-lab" AS (
  SELECT min(resting_heartrate) AS min_resting_heartrate,
         avg(resting_heartrate) AS avg_resting_heartrate,
         max(resting_heartrate) AS max_resting_heartrate,
         max(resting_heartrate) - min(resting_heartrate) AS resting_heartrate_change,
         min(active_heartrate) AS min_active_heartrate,
         avg(active_heartrate) AS avg_active_heartrate,
         max(active_heartrate) AS max_active_heartrate,
         max(active_heartrate) - min(active_heartrate) AS active_heartrate_change,
         min(bmi) AS min_bmi,
         avg(bmi) AS avg_bmi,
         max(bmi) AS max_bmi,
         max(bmi) - min(bmi) AS bmi_change,
         min(vo2) AS min_vo2,
         avg(vo2) AS avg_vo2,
         max(vo2) AS max_vo2,
         max(vo2) - min(vo2) AS vo2_change,
         min(workout_minutes) AS min_workout_minutes,
         avg(workout_minutes) AS avg_workout_minutes,
         max(workout_minutes) AS max_workout_minutes,
         max(workout_minutes) - min(workout_minutes) AS workout_minutes_change,
         CASE WHEN avg(steps) >= 10000 THEN 1 ELSE 0 END AS steps_10000
  FROM adsda.ht_daily_metrics
  GROUP BY device_id
)

In [0]:
%sql
SELECT * 
FROM adsda.ht_user_metrics_hs_lab

min_resting_heartrate avg_resting_heartrate max_resting_heartrate resting_heartrate_change min_active_heartrate avg_active_heartrate max_active_heartrate active_heartrate_change min_bmi avg_bmi max_bmi bmi_change min_vo2 avg_vo2 max_vo2 vo2_change min_workout_minutes avg_workout_minutes max_workout_minutes workout_minutes_change steps_10000 100.12190323385116 82.68379727873081 99.13828099149359 -0.9836222423575691 120.30779135992243 139.43487473206162 162.34782728523246 42.04003592531002 19.530815683202114 22.398063650890794 24.997081438324194 5.46626575512208 20.28334129074003 20.99401157735923 21.491176274984294 1.2078349842442648 1.0469302273257908 5.5026324666656405 9.325750004894761 8.27881977756897 0 52.71287564903336 77.73294228506452 97.93773119202254 45.22485554298918 109.04938651327842 127.05715346661702 146.86986814834722 37.8204816350688 22.63526877190205 25.150812654086298 27.93361433865782 5.298345566755767 24.929150796942785 25.52747526955064 26.231702127031934 1.3025513300891483 12.109056090051782 37.2167018100805 66.11591853237638 54.0068624423246 0 100.37366097806768 86.51162895591307 99.78933645583436 -0.5843245222333167 129.5517281106538 147.31573126952208 177.78314996951866 48.23142185886485 16.84474064400951 19.148256000462478 21.339664540140973 4.494923896131464 18.81713072211899 19.448406520026342 19.946846814093323 1.129716091974334 10.524606159589158 45.00008651086257 84.72505622254651 74.20045006295736 0 58.41880559722201 77.55054135762612 98.87532936858241 40.4565237713604 110.84551702943472 129.5770039396946 146.7386499653505 35.89313293591579 21.33582041108623 24.240375728856797 26.909149799032416 5.573329387946185 20.705859570960108 21.40130178285617 22.088526447872425 1.3826668769123174 11.402530422658407 37.886068725488464 68.11952462786931 56.7169942052109 0 49.81689070901131 68.93310580458204 92.68678860925904 42.86989790024773 116.68894230853843 136.50268661405897 162.3638297702582 45.67488746171978 27.4860355505171 30.726595797380472 33.87217279937333 6.38613724885623 28.102425066324887 28.855230169253637 29.7652881606717 1.6628630943468146 11.659988588325357 32.24198398599063 60.51374198315488 48.853753394829525 0 47.35569868189229 69.31244794850774 92.76249605385787 45.40679737196558 147.8761660158842 167.18585016710105 186.28361490243202 38.407448886547826 23.589288180958636 27.1326690342849 31.02646723458486 7.437179053626224 30.032977706533124 30.939205114246853 32.02682828439313 1.9938505778600089 0.8561243193074577 5.119426899323105 9.947677458057841 9.091553138750383 0 43.51461112051503 64.64397544858174 82.07381856112256 38.55920744060754 126.14263661366581 152.9654977304546 173.98748183735378 47.84484522368797 25.865189192070154 29.17716498363452 32.82229405815375 6.957104866083597 27.834198767673833 28.92795344089978 29.78220180035955 1.948003032685719 0.800631665598566 5.015081852287961 8.929350705708268 8.128719040109702 0 100.00309226714643 81.33282756113321 99.23741546147319 -0.7656768056732375 119.40382197892441 137.57131998347788 156.9670310686102 37.563209089685785 18.46292617354235 20.850071485672633 23.4081125062503 4.9451863327079515 22.030643616393576 22.564006304582495 23.21515783069579 1.1845142143022152 11.987667190790141 42.375521457262316 9.021257321662569 -2.966409869127572 0 40.714172943212596 64.79507042723496 82.96829328067244 42.254120337459845 116.41765662495979 139.39836367080545 157.39528580691194 40.97762918195215 28.32413026515547 31.386431213715433 34.813932679701345 6.489802414545874 28.2280680882799 29.096510773429188 29.984289559372073 1.7562214710921715 11.412819218675624 33.3298371336183 56.012582651278876 44.599763432603254 0 100.00362526100596 89.51117796589962 99.91629552898303 -0.08732973202293692 106.10261352131862 126.57048164605168 148.28320037619991 42.1805868548813 17.188042545277234 19.830753716401606 22.288022198086434 5.0999796528092 19.191314657468624 19.750462151303648 20.454506804319315 1.2631921468506917 12.979621455012532 43.305280

**Coursera Quiz:** How many users in `adsda.ht_user_metrics_hs_lab` take, on average, 10,000 steps per day?

In [0]:
%sql
SELECT steps_10000, count(*) FROM adsda.ht_user_metrics_hs_lab GROUP BY steps_10000

steps_10000 count(1) 1 1892 0 1108

-sandbox
## Exercise 2

In this exercise, you will split your data into a training set (`train_df`), validation set (`val_df`), and test set (`test_df`).

Fill in the blanks below to split your data.

<img alt="Hint" title="Hint" style="vertical-align: text-bottom; position: relative; height:1.75em; top:0.3em" src="https://files.training.databricks.com/static/images/icon-light-bulb.svg"/>&nbsp;**Hint:** Refer to the previous demo for guidance.

In [0]:
# ANSWER
from sklearn.model_selection import train_test_split

ht_user_metrics_pd_df = spark.table("adsda.ht_user_metrics_hs_lab").toPandas()

train_val_df, test_df = train_test_split(ht_user_metrics_pd_df, train_size=0.85, test_size=0.15, random_state=42)
train_df, val_df = train_test_split(train_val_df, train_size=0.7, test_size=0.3, random_state=42)

**Coursera Quiz:** How many rows are in the `val_df` DataFrame?

In [0]:
val_df.shape

Out[6]: (765, 21)

## Exercise 3

In this exercise, you will prepare your random forest classifier.

Fill in the blanks below to complete the task.

In [0]:
# ANSWER
from sklearn.ensemble import RandomForestClassifier

rfc = RandomForestClassifier(random_state=42)

## Exercise 4

In this exercise, you will create a hyperparameter grid to use during the grid search process.

Use the following hyperparameter values:

1. `max_depth`: 2, 3, 5, 8, 10, 15
1. `n_estimators`: 5, 10, 25, 50, 100, 250
1. `min_samples_split`: 2, 3, 4
1. `min_impurity_decrease`: 0.0, 0.01, 0.05

Fill in the blanks below to create the grid.

In [0]:
# ANSWER
parameter_grid = {
  "max_depth": [2, 3, 5, 8, 10, 15],
  "n_estimators": [5, 10, 25, 50, 100, 250],
  "min_samples_split": [2, 3, 4],
  "min_impurity_decrease": [0.0, 0.01, 0.05]
}

**Coursera Quiz**: How many total unique combinations of hyperparameters are there in `parameter_grid`?

In [0]:
len(parameter_grid["max_depth"]) * len(parameter_grid["n_estimators"]) * len(parameter_grid["min_samples_split"]) * len(parameter_grid["min_impurity_decrease"])

Out[9]: 324

## Exercise 5

In this exercise, you will create a predefined split for your training set and your validation set.

Fill in the blanks below to create the PredefinedSplit.

In [0]:
# ANSWER
from sklearn.model_selection import PredefinedSplit

# Create list of -1s for training set row or 0s for validation set row
split_index = [-1 if row in train_df.index else 0 for row in train_val_df.index]

# Create predefined split object
predefined_split = PredefinedSplit(test_fold=split_index)

**Coursera Quiz**: How many 0s are there in `split_index`?

In [0]:
split_index.count(0)

Out[11]: 765

## Exercise 6

In this exercise, you will create the grid search object that you will use to optimize your hyperparameter values.

Fill in the blanks below to create the object.

In [0]:
# ANSWER
from sklearn.model_selection import GridSearchCV

grid_search = GridSearchCV(estimator=rfc, cv=predefined_split, param_grid=parameter_grid)

## Exercise 7

In this exercise, you will fit the grid search process.

Fill in the blanks below to perform the grid search process.

In [0]:
# ANSWER
grid_search.fit(train_val_df.drop("steps_10000", axis=1), train_val_df["steps_10000"])

Out[13]: GridSearchCV(cv=PredefinedSplit(test_fold=array([-1, -1, ..., -1, 0])),
 error_score=nan,
 estimator=RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
 class_weight=None,
 criterion='gini', max_depth=None,
 max_features='auto',
 max_leaf_nodes=None,
 max_samples=None,
 min_impurity_decrease=0.0,
 min_impurity_split=None,
 min_samples_leaf=1,
 min_samples_split=2,
 min_weigh...0.0,
 n_estimators=100, n_jobs=None,
 oob_score=False, random_state=42,
 verbose=0, warm_start=False),
 iid='deprecated', n_jobs=None,
 param_grid={'max_depth': [2, 3, 5, 8, 10, 15],
 'min_impurity_decrease': [0.0, 0.01, 0.05],
 'min_samples_split': [2, 3, 4],
 'n_estimators': [5, 10, 25, 50, 100, 250]},
 pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
 scoring=None, verbose=0)

-sandbox
**Coursera Quiz**: How many unique models are being trained by the grid search process?

* 324
* 765
* 325
* 766

<img alt="Hint" title="Hint" style="vertical-align: text-bottom; position: relative; height:1.75em; top:0.3em" src="https://files.training.databricks.com/static/images/icon-light-bulb.svg"/>&nbsp;**Hint:** Consider the number of unique feature combinations and the final retraining of the model on the training *and* validation sets.

## Exercise 8

In this exercise, you will identify the optimal hyperparameter values.

Fill in the blanks below to indentify the optimal hyperparameter values.

In [0]:
# ANSWER
grid_search.best_params_

Out[14]: {'max_depth': 8,
 'min_impurity_decrease': 0.0,
 'min_samples_split': 3,
 'n_estimators': 25}

**Coursera Quiz:** What is the optimal hyperparameter value for `min_samples_split` according to the grid search process?

## Exercise 9

In this exercise, you will identify the validation accuracy that was achieved for the optimal hyperparameter values when trained on the training set.

Fill in the blanks below to identify the validation accuracy.

In [0]:
# ANSWER
grid_search.best_score_

Out[15]: 0.9882352941176471

**Coursera Quiz:** What is the best validation set accuracy?

## Exercise 10

In this exercise, you will identify the test accuracy achieved by the final, refit model.

Fill in the blanks below to identify the test accuracy.

In [0]:
# ANSWER
from sklearn.metrics import accuracy_score

accuracy_score(
  test_df["steps_10000"], 
  grid_search.predict(test_df.drop("steps_10000", axis=1))
)

Out[16]: 1.0

**Coursera Quiz:** What is the test set accuracy?

Congrats! That concludes our lesson on hyperparameter optimization!

Be sure to submit your quiz answers to Coursera, and join us in the next module to learn about how to improve the process even further using cross-validation.

-sandbox
&copy; 2020 Databricks, Inc. All rights reserved.<br/>
Apache, Apache Spark, Spark and the Spark logo are trademarks of the <a href="http://www.apache.org/">Apache Software Foundation</a>.<br/>
<br/>
<a href="https://databricks.com/privacy-policy">Privacy Policy</a> | <a href="https://databricks.com/terms-of-use">Terms of Use</a> | <a href="http://help.databricks.com/">Support</a>